src/librustc_mir/borrow_check/borrow_set.rs

```diff
@@ -107,7 +107,7 @@ impl LocalsStateAtExit {
             LocalsStateAtExit::AllAreInvalidated
         } else {
             let mut has_storage_dead = HasStorageDead(BitSet::new_empty(body.local_decls.len()));
-            has_storage_dead.visit_body(body);
+            has_storage_dead.visit_body(*body);
```

Comment:
Can you use `&body` instead of `*body` everywhere? I think it would've already coerced except `ReadOnlyBodyAndCache` is itself a reference internally so it's not passed by reference.

Although, I'm surprised `ReadOnlyBodyAndCache` exists, wouldn't `&BodyAndCache` also do the same thing?

Reply:
`ReadOnlyBodyAndCache` can only exist if the cache has already been precomputed. That guarantee doesn't hold with `&BodyAndCache`. That was the only reason a separate type was used. It would have been a lot easier if I could have just used `&BodyAndCache`.

Reply:
See https://github.com/rust-lang/rust/blob/a06baa56b95674fc626b3c3fd680d6a65357fe60/src/librustc/mir/cache.rs#L195-L198

Reply:
And the reason a variant with `&Body` and `&mut Cache` doesn't work is that there were some places that required such a struct to be `Copy`, which doesn't hold with the `&mut`. Though I can't remember which trait was causing that issue.